Case Insensitivity

davidsawyer opened the pull request:
`downcase` everything to prevent case mixups

Should fix https://github.com/iamvery/lita-reviewme/issues/61

davidsawyer committed:
Downcase everything to prevent case mixups

davidsawyer commented:
I'm just now realizing that this could pose an issue to people who already have non-lowercase users in their reviewer list. There are probably a few different ways of properly handling this, but I'm wondering if you have any initial thoughts on that issue.

iamvery commented:
Thanks for the PR! The implementation seems reasonable, but I understand the concern you're noting about existing reviewers. 🤔 I guess this implementation would make it impossible to remove any such existing reviewer, yes? I don't love the idea of having to run some maintenance task or build some list normalization routine into the handler. I guess another option would be to make reviewer removal case-insensitive rather than just normalized. Again that feels kind of "meh" given the add reviewer implementation is always downcase. What ideas do you have?

davidsawyer commented:
>I guess this implementation would make it impossible to remove any such existing reviewer, yes?

Correct, it would be impossible to remove any reviewer who has a non-zero amount of uppercase characters in their username.

>What ideas do you have?

I definitely get the hesitation to run maintenance tasks in general, but I'm seeing in the [Lita docs](https://docs.lita.io/plugin-authoring/events/) that they recommend hooking into the `:loaded` event to run migrations on Redis data, so that might be a reasonable approach. It could just loop through all values in the list of reviewers and ensure all of them are lowercase, and it would only happen once on startup whenever Lita is upgraded and/or restarted.

Let me know what you think about this approach!

iamvery commented:
That sounds decent to me :). The whole thing isn't ideal, so this is just a interest paid on tech debt. Feel free to write up a change, and we can discuss. Thanks for jumping in! I hope you've found the lib useful ❤️💙💚💛💜

davidsawyer commented:
Will do, thanks!

davidsawyer committed:
Fix typo

davidsawyer committed:
Make all usernames lowercase on startup

to make sure we aren't leaving behind any non-lowercase usernames with our new case insensitivity feature.

davidsawyer commented:
Just pushed up that change. Still not in love with this approach, but if we don't want to be `downcase`ing an entire list every single time the user goes to remove a reviewer, then this approach feels like a relatively sensible way forward.

iamvery commented on lib/lita/handlers/reviewme.rb:
```diff
@@ -116,6 +118,19 @@ def mention_reviewer(response, room: get_room(response))
+      # need to migrate all reviewer lists to make sure all usernames are lowercase, or otherwise
+      # users would not be able to remove usernames that already had uppercase characters in them.
+      def ensure_usernames_are_lowercase(payload)
+        lists = redis.keys("*")
```
It's been awhile since I've worked in redis, but this seems a little greedy. Is the `redis` instance already namespaced to only to this handler or something?

davidsawyer replied:
Yep, exactly. So this would only return keys that begin with `lita:handlers:reviewme:`. You think it's worth a comment above this line noting that nuance?

iamvery replied:
Yes, that sounds useful :)

davidsawyer replied:
Alright, just pushed that up!

iamvery replied:
Hurray, thank you! It'll be next week before I can closely look at this. I want to think about it all a little more before merge. Thank you so much for all your work on it! 🎉

davidsawyer committed:
Add clarifying comment

iamvery commented on lib/lita/handlers/reviewme.rb:
```diff
@@ -116,6 +118,19 @@ def mention_reviewer(response, room: get_room(response))
+      # need to migrate all reviewer lists to make sure all usernames are lowercase, or otherwise
+      # users would not be able to remove usernames that already had uppercase characters in them.
+      def ensure_usernames_are_lowercase(payload)
+        lists = redis.keys("*") # get all reviewme namespaced keys
```
Sorry for the hilariously late comment here! 😓

I think what kept holding me up on this is how generalized the effect is. That is, this will have the affect of downcasing everything in the lita-reviewme namespace. Unfortunately that seems like a result of not nested "rooms" under a common key. There was nothing you could use to target the lists 😖.

I also don't love how this code has to sort of live on forever to ensure that really hold clients get correctly migrated when they upgrade. Oof. There's a part of make that wants to just make this a breaking change and let people rebuild their reviewer lists... Is there a better way?

davidsawyer replied:
Hmm, not that I can think of right now. I'll try to think about this more, but I'd be happy to hear your or (anyone else's) suggestions on how to move forward here.

iamvery replied:
I feel really crummy about holding this up for so long, but I also feel a little crummy about adding ephemeral behavior without having a clear understanding of _when_ it can go away :/. I suppose we could build a little task or something that can be run to migrate existing data? Not sure how that fits in with Lita, but maybe as simple as a rake task?

davidsawyer replied:
Yeah, it'd be much cleaner if we could run a one-off migration task when the user upgraded from a previous version to the version that would contain this change.

iamvery replied:
It's been awhile since I've managed a Lita instance with this plugin. Any intuition if such a thing is doable?

davidsawyer replied:
Dang, I can't find much on it after a quick search. I'm honestly okay just letting this PR sit until someone comes along with a better solution.

iamvery replied:
Well thank you so much for you patience and understanding! I should probably be the one finding time to do it 😅 ❤️💙💚💛💜
